kratos/containers/distributed_csr_matrix.h

```diff
@@ -456,7 +456,48 @@ class DistributedCsrMatrix final
         return tmp;
     }
 
+    void ApplyHomogeneousDirichlet(const DistributedSystemVector<TDataType,TIndexType>& free_dofs_vector, 
+                                    const TDataType DiagonalValue, 
+                                    DistributedSystemVector<TDataType,TIndexType>& rRHS)
+    {
+        KRATOS_TRY
+        KRATOS_ERROR_IF(local_size1() != free_dofs_vector.LocalSize() ) << "ApplyHomogeneousDirichlet: mismatch between row sizes : " << local_size1()  
+            << " and free_dofs_vector size " << free_dofs_vector.LocalSize() << std::endl;
+
 
+        //diagonal block
+        GetDiagonalBlock().ApplyHomogeneousDirichlet(free_dofs_vector.GetLocalData(), DiagonalValue, rRHS.GetLocalData());
+
+        //off diagonal block
+        auto off_diag_free = mpVectorImporter->ImportData(free_dofs_vector); //obtain the nonlocal components of the free_dofs_vector
+        if(GetOffDiagonalBlock().nnz() != 0)
+        {
+            KRATOS_ERROR_IF(GetOffDiagonalBlock().size1() != free_dofs_vector.LocalSize() ) << "ApplyHomogeneousDirichlet - OffDiagonalBlock: mismatch between row sizes : " << GetOffDiagonalBlock().size1()  
+                << " and free_dofs_vector size " << free_dofs_vector.LocalSize() << std::endl;
+            KRATOS_ERROR_IF(GetOffDiagonalBlock().size2() != off_diag_free.size() ) << "ApplyHomogeneousDirichlet - OffDiagonalBlock: mismatch between col sizes : " << GetOffDiagonalBlock().size2()  
+                << " and free_dofs_vector size " << off_diag_free.size() << std::endl;
+            IndexPartition<IndexType>(GetDiagonalBlock().size1()).for_each( [&](IndexType i){
+
+                IndexType row_begin = GetOffDiagonalBlock().index1_data()[i];
+                IndexType row_end   = GetOffDiagonalBlock().index1_data()[i+1];
+                if(std::abs(free_dofs_vector.GetLocalData()[i] - 1.0) < 1e-10) //row corresponding to free dofs NOTE that we check if it is approx zero
```

Comment:
Please, follow the style as previously stated

Reply:
no, i prefer the style as it is now

Reply:
that was too harsh, sorry.

what i mean is that here we have two quite large blocks. i prefer to have them separated more clearly